### ISSUE TITLE
Conflict between parallel password and fingerprint authentication

### ISSUE THREAD

**Issue by rickyelopez** [hyprwm/hyprlock#608]
### Regression?

No

### Hyprlock Info and Version

Hyprlock version 0.6.0
<details> <summary>Hyprlock config</summary>

```sh
general {
  grace = 1
  no_fade_in = false
  no_fade_out = false
}

auth {
  fingerprint {
    enabled = true
  }
}

background {
  monitor = 
  path = screenshot
  color = rgba(25, 20, 20, 1.0)
  blur_passes = 2
}

input-field {
  monitor =
  size = 200, 50
  outline_thickness = 3
  dots_size = 0.11
  fade_timeout = 2000

  position = 0, -300
  halign = center
  valign = center
}

label {
  monitor =
  text = $TIME
  color = rgb(100, 100, 100)
  font_size = 50
  font_family = IBM Plex

  position = 0, 250
  halign = center
  valign = center
}
```

</details>

### Compositor Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.2.6
 hyprcursor 0.1.10
 hyprgraphics 0.1.1


flags set:
debug


System Information:
System name: Linux
Node name: expedition
Release: 6.12.4-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Mon, 09 Dec 2024 14:31:57 +0000


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation CometLake-H GT2 [UHD Graphics] [8086:9bc4] (rev 05) (prog-if 00 [VGA controller])
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation TU117M [GeForce GTX 1650 Mobile / Max-Q] [10de:1f99] (rev a1) (prog-if 00 [VGA controller])
NVRM version: NVIDIA UNIX x86_64 Kernel Module  565.77  Wed Nov 27 23:33:08 UTC 2024


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo
```

</details>

### Description

Really cool that support for fingerprint auth was added! I'm not sure if this is a result of the weird fprint packages I have installed to support my weird fingerprint sensor or if it's a bug in hyprlock.

fprint appears to constantly retry fingerprint verificationr regardless of whether I actually have my finger on the sensor or not. When I don't have my finger on the sensor, this means the fingerprint verification fails every couple of seconds, over and over again.

Every time fingerprint auth fails, it clears the password that I was in the middle of typing, which means that unless I wait for the attempt to fail and then type my password _really_ fast, I can't actually get in with password auth.

This is with
fprintd: 1.94.4-1
libfprint-tod-git: 1.94.7+tod1.r46.g5d0de75-1
libfprint-2-tod1-goodix: 0.0.9.r3.g5f05f60-1

Not sure what other info might be relevant but I'm happy to provide anything I can

### How to reproduce

Enable fingerprint auth and launch hyprlock

### Crash reports, logs, images, videos

<details> <summary>hyprlock log with failed verification attempts</summary>
[LOG]   | got iface: wl_seat v9
[LOG]    > Bound to wl_seat v9
[LOG]   | got iface: wl_data_device_manager v3
[LOG]   | got iface: wl_compositor v6
[LOG]    > Bound to wl_compositor v6
[LOG]   | got iface: wl_subcompositor v1
[LOG]   | got iface: wl_shm v1
[LOG]   | got iface: wp_viewporter v1
[LOG]    > Bound to wp_viewporter v1
[LOG]   | got iface: wp_tearing_control_manager_v1 v1
[LOG]   | got iface: wp_fractional_scale_manager_v1 v1
[LOG]    > Bound to wp_fractional_scale_manager_v1 v1
[LOG]   | got iface: zxdg_output_manager_v1 v3
[LOG]   | got iface: wp_cursor_shape_manager_v1 v1
[LOG]    > Bound to wp_cursor_shape_manager_v1 v1
[LOG]   | got iface: zwp_idle_inhibit_manager_v1 v1
[LOG]   | got iface: zwp_relative_pointer_manager_v1 v1
[LOG]   | got iface: zxdg_decoration_manager_v1 v1
[LOG]   | got iface: wp_alpha_modifier_v1 v1
[LOG]   | got iface: zwlr_gamma_control_manager_v1 v1
[LOG]   | got iface: ext_foreign_toplevel_list_v1 v1
[LOG]   | got iface: zwp_pointer_gestures_v1 v3
[LOG]   | got iface: zwlr_foreign_toplevel_manager_v1 v3
[LOG]   | got iface: zwp_keyboard_shortcuts_inhibit_manager_v1 v1
[LOG]   | got iface: zwp_text_input_manager_v1 v1
[LOG]   | got iface: zwp_text_input_manager_v3 v1
[LOG]   | got iface: zwp_pointer_constraints_v1 v1
[LOG]   | got iface: zwlr_output_power_manager_v1 v1
[LOG]   | got iface: xdg_activation_v1 v1
[LOG]   | got iface: ext_idle_notifier_v1 v1
[LOG]   | got iface: ext_session_lock_manager_v1 v1
[LOG]    > Bound to ext_session_lock_manager_v1 v1
[LOG]   | got iface: zwp_input_method_manager_v2 v1
[LOG]   | got iface: zwp_virtual_keyboard_manager_v1 v1
[LOG]   | got iface: zwlr_virtual_pointer_manager_v1 v2
[LOG]   | got iface: zwlr_output_manager_v1 v4
[LOG]   | got iface: org_kde_kwin_server_decoration_manager v1
[LOG]   | got iface: hyprland_focus_grab_manager_v1 v1
[LOG]   | got iface: zwp_tablet_manager_v2 v1
[LOG]   | got iface: zwlr_layer_shell_v1 v5
[LOG]   | got iface: wp_presentation v1
[LOG]   | got iface: xdg_wm_base v6
[LOG]   | got iface: zwlr_data_control_manager_v1 v2
[LOG]   | got iface: zwp_primary_selection_device_manager_v1 v1
[LOG]   | got iface: xwayland_shell_v1 v1
[LOG]   | got iface: zwlr_screencopy_manager_v1 v3
[LOG]    > Bound to zwlr_screencopy_manager_v1 v3
[LOG]   | got iface: hyprland_toplevel_export_manager_v1 v2
[LOG]   | got iface: hyprland_global_shortcuts_manager_v1 v1
[LOG]   | got iface: xdg_dialog_v1 v1
[LOG]   | got iface: wp_single_pixel_buffer_manager_v1 v1
[LOG]   | got iface: wp_security_context_manager_v1 v1
[LOG]   | got iface: hyprland_ctm_control_manager_v1 v1
[LOG]   | got iface: wp_drm_lease_device_v1 v1
[LOG]   | got iface: wp_linux_drm_syncobj_manager_v1 v1
[LOG]   | got iface: wl_drm v2
[LOG]   | got iface: zwp_linux_dmabuf_v1 v5
[LOG]    > Bound to zwp_linux_dmabuf_v1 v5
[LOG]   | got iface: wl_output v4
[LOG]    > Bound to wl_output v4
[LOG] [core] dmabufFeedbackMainDevice
[LOG] output 51 name eDP-1
[LOG] output 51 description LG Display 0x0637  (eDP-1)
[LOG] output 51 make LG Display model 0x0637
[LOG] output 51 done
[LOG] Running on Hyprland
[LOG] Locking session
[LOG] eglQueryDmaBufModifiersEXT found 4 mods
[LOG] bo chose modifier 0
[LOG] fprint: using device path /net/reactivated/Fprint/Device/0
[LOG] Got fractional scale: 1
[LOG] Got fractional scale: 1
[LOG] configure with serial 127089
[LOG] Configuring surface for logical [Vector2D: x: 1920, y: 1080] and pixel [Vector2D: x: 1920, y: 1080]
[LOG] Got fractional scale: 1
[LOG] Got dma frame with size [Vector2D: x: 1920, y: 1080]
[LOG] onLockLocked called
[LOG] fprint: claimed device
[LOG] fprint: started verifying
[LOG] fprint: finger selected: right-index-finger
[LOG] fprint: handling status verify-retry-scan
[LOG] Failed attempts: 1
[LOG] fprint: handling status verify-retry-scan
[LOG] Failed attempts: 2
[LOG] fprint: handling status verify-retry-scan
[LOG] Failed attempts: 3
[LOG] fprint: handling status verify-retry-scan
[LOG] Failed attempts: 4
[LOG] fprint: handling status verify-no-match
[LOG] fprint: stopped verification
[LOG] Failed attempts: 5
[LOG] fprint: started verifying
[LOG] fprint: finger selected: right-index-finger
[LOG] fprint: handling status verify-match
[LOG] fprint: stopped verification
[LOG] Unlocking session
[LOG] Unlocked, exiting!
[LOG] fprint: released device
[LOG] Reached the end, exiting
</details>

**Comment by PaideiaDilemma** [comment]
Well ups, i added input field clearing, when a fingerprint attempt was made, thinking it was reasonable behaviour.

> fprint appears to constantly retry fingerprint verificationr regardless of whether I actually have my finger on the sensor or not. When I don't have my finger on the sensor, this means the fingerprint verification fails every couple of seconds, over and over again.

Honestly that sounds like a diver issue to me.

**Comment by rickyelopez** [comment]
why clear the field when a fingerprint attempt is made? Shouldn't the two methods of authentication be independant?

**Comment by PaideiaDilemma** [comment]
because the. it will show fail_text...

**Comment by rickyelopez** [comment]
Oh that makes sense. Could it be a config option? eg. "clear_input_on_fprint_fail" or something? I can see the benefit of the way that you have it, but unfortunately I think this type of fingerprint sensor (TOD, touch-only device?) might be susceptible to this issue?

I cleaned up my config quite a bit after all the debugging it took to get the fingerprint sensor working at all. I am now down to just:

fprintd: 1.94.4-1
libfprint-goodix-521d: 1.94.1-8

I don't think the offshoot libfprint I'm using is what has changed this behavior, I'm leaning more towards the touch-only devices requiring it...

**Comment by rickyelopez** [comment]
After looking at the code for a bit, would this be solved if the `MATCH_RETRY` branch just didn't result in calling `enqueueFail`? It's not really a failure, I don't think. It would still be nice to have the option to print a message that the retry happened, but does it need to be considered a failure?

**Comment by rickyelopez** [comment]
Should be addressed by #613
